source/extensions/tracers/datadog/event_scheduler.cc

```diff
@@ -0,0 +1,62 @@
+#include "source/extensions/tracers/datadog/event_scheduler.h"
+
+#include <cassert>
+#include <datadog/json.hpp>
+#include <memory>
+#include <utility>
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+
+EventScheduler::EventScheduler(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+EventScheduler::~EventScheduler() {
+  for (const auto& timer : timers_) {
+    timer->disableTimer();
+  }
+}
+
+EventScheduler::Cancel
+EventScheduler::schedule_recurring_event(std::chrono::steady_clock::duration interval,
+                                         std::function<void()> callback) {
+  // Yes, a shared pointer to a pointer.
+  auto self = std::make_shared<Event::Timer*>();
+
+  Event::TimerPtr timer = dispatcher_.createTimer([self, interval, callback = std::move(callback)] {
+    (**self).enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(interval));
+    callback();
+  });
+
+  Event::Timer* timer_raw = timer.get();
+  *self = timer_raw;
+
+  timers_.emplace_back(std::move(timer));
+
+  timer_raw->enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(interval));
+
+  return [this, timer = timer_raw]() mutable {
+    if (!timer) {
+      return; // idempotent
+    }
+
+    timer->disableTimer();
+    auto found = std::find_if(timers_.begin(), timers_.end(),
+                              [&](const auto& element) { return element.get() == timer; });
+    assert(found != timers_.end());
+    timers_.erase(found);
+    timer = nullptr;
+  };
+}
```

Comment:
The contract of the function says that use of the cancellation function after this object is destroyed results in undefined behavior. Mean-spirited, perhaps, but I don't think that it can happen in any case.

What kind of guard might be appropriate?

My thinking is that the `assert` does the job of crashing during testing should we mistakenly break the contract in the future. I should probably replace it with one of Envoy's assert macros instead.

Reply:
Any possible runtime error or crash is dangerous for Envoy (as gateway/lb or sidecar) and we will try to ensure it will not happend.
But I have no much background knowledge about these scheduler, so may be I am wrong.

A guard could be a simple shared_pointer of bool. We can catch this pointer in every async callbck. And when the object is destroyed, we can set the value of bool to false. And the async callback run accutually logic only when the value of bool is still be true.

Reply:
That would work.

Then we'd have to test it, though, and we'd be testing a code path that "shall not" actually happen. I think that a `RELEASE_ASSERT` suffices here. Let me know if you disagree.

Reply:
I am ok if you ensure that "that "shall not" actually happen". 😸

Reply:
I prefer to leave it as undefined behavior, though I appreciate your instinct of avoiding footguns.